src/btree/iter.rs

```diff
@@ -68,232 +94,178 @@ impl<'a> BTreeIterator<'a> {
-			let commit_overlay = self.commit_overlay.read();
-			self.from_seek |= self.direction == IterDirection::Backward;
-			let next_commit_overlay = commit_overlay
-				.get(col as usize)
-				.and_then(|o| o.btree_next(&self.last_key, self.from_seek));
-			let log = self.log.read();
-			let record_id = log.last_record_id(self.col);
-			// No consistency over iteration, allows dropping lock to overlay.
-			drop(commit_overlay);
-			if record_id != self.iter.1.record_id {
-				self.pending_next_backend = None;
+			if let Ok(result) = self.iter_inner(IterDirection::Forward) {
+				return result
 			}
-			let next_backend = if self.pending_next_backend.is_some() &&
-				self.direction == IterDirection::Forward
-			{
-				self.pending_next_backend.take().expect("Checked above")
-			} else {
-				self.next_backend(record_id, self.table, &*log)?
-			};
-
-			match (next_commit_overlay, next_backend) {
-				(Some((commit_key, commit_value)), Some((backend_key, backend_value))) =>
-					match (commit_key.cmp(&backend_key), commit_value) {
-						(std::cmp::Ordering::Less, Some(value)) => {
-							self.last_key = Some(commit_key.clone());
-							self.from_seek = false;
-							self.pending_next_backend = Some(Some((backend_key, backend_value)));
-							return Ok(Some((commit_key, value)))
-						},
-						(std::cmp::Ordering::Less, None) => {
-							self.last_key = Some(commit_key);
-							self.from_seek = false;
-							self.pending_next_backend = Some(Some((backend_key, backend_value)));
-						},
-						(std::cmp::Ordering::Greater, _) => {
-							self.last_key = Some(backend_key.clone());
-							return Ok(Some((backend_key, backend_value)))
-						},
-						(std::cmp::Ordering::Equal, Some(value)) => {
-							self.last_key = Some(commit_key);
-							self.from_seek = false;
-							return Ok(Some((backend_key, value)))
-						},
-						(std::cmp::Ordering::Equal, None) => {
-							self.last_key = Some(commit_key);
-							self.from_seek = false;
-						},
-					},
-				(Some((commit_key, Some(commit_value))), None) => {
-					self.last_key = Some(commit_key.clone());
-					self.from_seek = false;
-					self.pending_next_backend = Some(None);
-					return Ok(Some((commit_key, commit_value)))
-				},
-				(Some((commit_key, None)), None) => {
-					self.last_key = Some(commit_key);
-					self.from_seek = false;
-					self.pending_next_backend = Some(None);
-				},
-				(None, Some((backend_key, backend_value))) => {
-					self.last_key = Some(backend_key.clone());
-					return Ok(Some((backend_key, backend_value)))
-				},
-				(None, None) => {
-					self.pending_next_backend = Some(None);
-					return Ok(None)
-				},
+		}
+	}
+
+	pub fn prev(&mut self) -> IterResult {
+		loop {
+			if let Ok(result) = self.iter_inner(IterDirection::Backward) {
+				return result
 			}
 		}
 	}
 
-	pub fn prev(&mut self) -> Result<Option<(Vec<u8>, Vec<u8>)>> {
+	fn iter_inner(&mut self, direction: IterDirection) -> std::result::Result<IterResult, ()> {
```

Comment:
I believe it would make the code more clear to move the `loop` from `prev` and `next` here and have explicit `continue` call instead of returning `Err`. It took me some time to figure out that returning `Err` is not a falure but a retry... But it's code style so feel free to ignore it...

Reply:
yes, @i1i1 did write it this way too, I guess that's better, I did switch.
Warning I don't release the log lock anymore.